### PR TITLE
Add CombineZernikes...Tasks.

### PR DESCRIPTION
**Pull request by jbkalmbach**

### CODE REVIEW

File: doc/content.rst [hunk 1]
```diff
@@ -99,3 +99,8 @@ This module has the tasks to run WEP as a pipeline with Gen 3 LSST DM middleware
 * **EstimateZernikesCwfsTaskConfig**: Configuration setup for EstimateZernikesCwfsTask.
 * **EstimateZernikesCwfsTask**: Gen 3 middleware task to take exposures and donut source catalogs and calculate Zenikes coefficients for each CCD when running on corner wave front sensors (CWFS). Saves Zernike coefficients and associated DonutStamps to Gen 3 repository.
 * **RefCatalogInterface**: Tools to pick out the pieces of a reference catalog in the Gen3 Butler that cover the sky area of a pointing.
+* **CombineZernikesBase**: Base class for CombineZernikes tasks that combine the Zernike coefficients from multiple donuts on a detector into a single set of coefficients for the detector.
```
Comment (teweitsai) [inline doc/content.rst]: **CombineZernikesBase** --> **CombineZernikesBaseTask**

File: doc/content.rst [hunk 2]
```diff
@@ -99,3 +99,8 @@ This module has the tasks to run WEP as a pipeline with Gen 3 LSST DM middleware
...
+* **CombineZernikesConfig**: Configuration setup for CombineZernikesBase.
```
Comment (teweitsai) [inline doc/content.rst]: Configuration setup for CombineZernikesBase --> Configuration setup for CombineZernikesBaseTask.

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        self.log = logging.getLogger(type(self).__name__)
+
+    def run(self, zernikeArray):
+        """
+        Combine the zernikes from the input array of Zernike
+        coefficients from each individual donut pair.
+
+        Parameters
+        ----------
+        zernikeArray: numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: zernikeArray: numpy ndarray --> zernikeArray : `numpy.ndarray`

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            The full set of zernike coefficients for each pair
+            of donuts on the CCD. Each row of the array should
+            be the set of Zernike coefficients for a single
+            donut pair.
+
+        Returns
+        -------
+        struct: `lsst.pipe.base.Struct`
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: struct: `lsst.pipe.base.Struct` --> struct : `lsst.pipe.base.Struct`

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            The struct contains the following data:
+
+            - combinedZernikes: numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: combinedZernikes: numpy ndarray --> combinedZernikes : `numpy.ndarray`

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                The final combined Zernike coefficients from the CCD.
+            - combineFlags: numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: combineFlags: numpy ndarray --> combineFlags : `numpy.ndarray`

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                Flag indicating a particular set of Zernike
+                coefficients was not used in the final estimate.
+                A value of 1 means the coefficients from that row
+                in the input `zernikeArray` were not used.
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py:75]: If used, what is the value? 0?

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    @abc.abstractmethod
+    def combineZernikes(self, zernikeArray):
+        """
+        Class specific algorithm to combine the Zernike
+        coefficients from each individual donut pair into
+        a single set of coefficients for the detector.
+
+        Parameters
+        ----------
+        zernikeArray: numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: zernikeArray: numpy ndarray --> zernikeArray : `numpy.ndarray`

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            of donuts on the CCD. Each row of the array should
+            be the set of Zernike coefficients for a single
+            donut pair.
+
+        Returns
+        -------
+        numpy ndarray
+            The final combined Zernike coefficients from the CCD.
+        numpy ndarray
+            A binary array where a value of 1 in any index indicates
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: If used, what is the value? 0?

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        ----------
+        zernikeArray: numpy ndarray
+            The full set of zernike coefficients for each pair
+            of donuts on the CCD. Each row of the array should
+            be the set of Zernike coefficients for a single
+            donut pair.
+
+        Returns
+        -------
+        numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: numpy ndarray --> `numpy.ndarray`

File: python/lsst/ts/wep/task/CombineZernikesBase.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+            The final combined Zernike coefficients from the CCD.
+        numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesBase.py]: numpy ndarray --> `numpy.ndarray`

File: python/lsst/ts/wep/task/CombineZernikesMeanTask.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+        Returns
+        -------
+        numpy ndarray
+            The final combined Zernike coefficients from the CCD.
+        numpy ndarray
+            A binary array where a value of 1 in any index indicates
+            that the row in the zernikeArray was not used
+            in the final combination.
+        """
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesMeanTask.py]: I think you could remove the doc string here. This function here inherits from the parent. If someone use the `help()` in the iPython or Jupiter notebook, he/she will get the doc string from the parent directly. You could try this to see what will happen. Thanks!

File: python/lsst/ts/wep/task/CombineZernikesSigmaClipTask.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+            donut pair.
+
+        Returns
+        -------
+        numpy ndarray
+            The final combined Zernike coefficients from the CCD.
+        numpy ndarray
+            The indices of the rows in the zernikeArray that were used
+            in the final combination.
+        """
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/CombineZernikesSigmaClipTask.py]: I think you could remove the doc string here. This function here inherits from the parent. If someone use the `help()` in the iPython or Jupiter notebook, he/she will get the doc string from the parent directly. You could try this to see what will happen. Thanks!

File: python/lsst/ts/wep/task/EstimateZernikesBase.py [hunk 14]
```diff
@@ -444,10 +458,16 @@ def combineZernikes(self, zernikeArray):
 
         Returns
         -------
-        finalZernikes: numpy ndarray
-            The final combined Zernike coefficients from the CCD.
+        struct: `lsst.pipe.base.Struct`
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesBase.py]: struct: `lsst.pipe.base.Struct` --> struct : `lsst.pipe.base.Struct`

File: python/lsst/ts/wep/task/EstimateZernikesBase.py [hunk 15]
```diff
@@ -444,10 +458,16 @@ def combineZernikes(self, zernikeArray):
...
+            The struct contains the following data:
+
+            - combinedZernikes: numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesBase.py]: combinedZernikes: numpy ndarray --> combinedZernikes : `numpy.ndarray`

File: python/lsst/ts/wep/task/EstimateZernikesBase.py [hunk 16]
```diff
@@ -444,10 +458,16 @@ def combineZernikes(self, zernikeArray):
...
+                The final combined Zernike coefficients from the CCD.
+            - combineFlags: numpy ndarray
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesBase.py]: combineFlags: numpy ndarray --> combineFlags : `numpy.ndarray`

File: python/lsst/ts/wep/task/EstimateZernikesBase.py [hunk 17]
```diff
@@ -444,10 +458,16 @@ def combineZernikes(self, zernikeArray):
...
+                Flag indicating a particular set of Zernike
+                coefficients was not used in the final estimate.
+                A value of 1 means the coefficients from that row
+                in the input `zernikeArray` were not used.
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesBase.py:472]: If used, what is the value? 0?

File: python/lsst/ts/wep/task/EstimateZernikesCwfsTask.py [hunk 18]
```diff
@@ -211,7 +211,8 @@ def run(
 
         # Estimate Zernikes from collection of stamps
         zernikeCoeffsRaw = self.estimateZernikes(donutStampsExtra, donutStampsIntra)
-        zernikeCoeffsAvg = self.combineZernikes(zernikeCoeffsRaw)
+        zernikeCoeffsCombined = self.getCombinedZernikes(zernikeCoeffsRaw)
+        zernikeCoeffsAvg = zernikeCoeffsCombined.combinedZernikes
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesCwfsTask.py]: Just want to make sure you noticed that these two variables (`zernikeCoeffsAvg` and `zernikeCoeffsCombined.combinedZernikes`) point to the same object right now. Usually, I will use the `copy()` to make the `zernikeCoeffsAvg` to point to another object with the same values. But I noticed that when you returned the value, you used the `np.array(zernikeCoeffsAvg)`, which will point to another object with the same value. You could use the `id()` to do some experiments to see what I mean here.

For the case here, it is fine. However, this implementation is a little risky for other developers if they did not notice this in the future. If it is me, I will do something like this:

```
zernikeCoeffsCombined = self.getCombinedZernikes(zernikeCoeffsRaw)

return pipeBase.Struct(
            outputZernikesAvg=np.array(zernikeCoeffsCombined.combinedZernikes),
            outputZernikesRaw=np.array(zernikeCoeffsRaw),
            donutStampsExtra=donutStampsExtra,
            donutStampsIntra=donutStampsIntra,
        )
```

File: python/lsst/ts/wep/task/EstimateZernikesScienceSensorTask.py [hunk 19]
```diff
@@ -200,7 +200,8 @@ def run(
 
         # Estimate Zernikes from collection of stamps
         zernikeCoeffsRaw = self.estimateZernikes(donutStampsExtra, donutStampsIntra)
-        zernikeCoeffsAvg = self.combineZernikes(zernikeCoeffsRaw)
+        zernikeCoeffsCombined = self.getCombinedZernikes(zernikeCoeffsRaw)
+        zernikeCoeffsAvg = zernikeCoeffsCombined.combinedZernikes
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesScienceSensorTask.py]: Same comment in `EstimateZernikesCwfsTask.py`.

**Review by teweitsai — APPROVED** [review]
Looks good to me. Thanks for the update! Please consider the improve of document and do some experiments of `id()`. Thanks!

File: python/lsst/ts/wep/task/CombineZernikesMeanTask.py (hunk 12 above)
Comment (jbkalmbach) [reply python/lsst/ts/wep/task/CombineZernikesMeanTask.py]: Oh, cool. I just checked this out and it works. Thanks, I didn't know this.

**Comment by jbkalmbach** [comment]
Waiting to merge until Tiago runs some tests on MTAOS pipelines.

**Pull request merged**